File: src/services/fiat-fx.ts
/**
 * Fiat FX rates fetched directly on the client (no server proxy) from the
 * user-configurable endpoint (`ServiceEndpoints.fiatRatesURL`, default
 * Frankfurter v2 — FOSS, self-hostable, ~160 currencies incl. VND).
 *
 * The endpoint must return USD-based rates in one of two shapes (`normalizeRates`
 * accepts both, so providers are swappable):
 *
 *   Frankfurter v2 (array):  [{ base:'USD', quote:'EUR', rate:0.92 }, …]
 *   open.er-api / v1 (object): { rates: { EUR: 0.92, … } }
 *
 * Either way `rate` is X per 1 USD — exactly the USD→fiat multiplier the app
 * needs. The provider decides which currencies exist, so the displayed list is
 * whatever it returns (plus Chainlink + USD).
 *
 * Cached in-memory (6h) + persisted for offline/first-paint; the cache is keyed
 * by endpoint URL so changing the endpoint refetches.
 */
import AsyncStorage from '@react-native-async-storage/async-storage';
import { getFiatRatesURL } from '@/services/storage';

/** Normalize a provider response (array or `{rates}` object) to `{ USD:1, X:rate }`, or null. */
export function normalizeRates(data: unknown): Record<string, number> | null {
  const out: Record<string, number> = { USD: 1 };
  if (Array.isArray(data)) {
    // Frankfurter v2: [{ base, quote, rate }] (base must be USD)
    for (const row of data as Array<{ quote?: string; rate?: unknown }>) {
      const code = String(row?.quote ?? '').toUpperCase();
      const n = Number(row?.rate);
      if (code && isFinite(n) && n > 0) out[code] = n;
    }
  } else if (data && typeof (data as any).rates === 'object') {
    // open.er-api / Frankfurter v1: { rates: { CODE: rate } }
    for (const [k, v] of Object.entries((data as any).rates as Record<string, unknown>)) {
      const n = Number(v);
      if (isFinite(n) && n > 0) out[k.toUpperCase()] = n;
    }
  }
  return Object.keys(out).length > 1 ? out : null;
}

const CACHE_KEY = 'vela.fxRates.v1';
// ECB / Frankfurter rates update once per business day (~16:00 CET), so a long
// cache is plenty — we refetch a few times a day, not on every screen open.
const TTL = 6 * 60 * 60 * 1000; // 6h

interface FxCache { url: string; rates: Record<string, number>; at: number }

let _cache: FxCache | null = null;
let _inflight: Promise<Record<string, number>> | null = null;

/**
 * USD-based rate map for every currency the configured endpoint returns
 * (`{ USD: 1, EUR: 0.92, … }`). Cached 1h; falls back to the persisted map.
 */
export async function fetchFxRates(): Promise<Record<string, number>> {
  const url = getFiatRatesURL();
  if (_cache && _cache.url === url && Date.now() - _cache.at < TTL) return _cache.rates;
  if (_inflight) return _inflight;

  _inflight = (async () => {
    try {
      const res = await fetch(url);
      if (res.ok) {
        const rates = normalizeRates(await res.json());
        if (rates) {
          _cache = { url, rates, at: Date.now() };
          AsyncStorage.setItem(CACHE_KEY, JSON.stringify(_cache)).catch(() => {});
          console.log(`[FX] ${Object.keys(rates).length} rates from ${url}`);
          return rates;
        }
      }
      console.warn(`[FX] bad response from ${url} (HTTP ${res.status})`);
    } catch (err) {
      console.warn(`[FX] fetch failed: ${err instanceof Error ? err.message : String(err)}`);
    }
    return loadPersisted();
  })();

  try { return await _inflight; } finally { _inflight = null; }
}

/** USD → `code` multiplier from the configured endpoint, or null if unavailable. */
export async function getFxRate(code: string): Promise<number | null> {
  const c = code.toUpperCase();
  if (c === 'USD') return 1;
  const rates = await fetchFxRates();
  const r = rates[c];
  return r && r > 0 ? r : null;
}

/** Currency codes the configured endpoint can price (drives the offered list). */
export async function getSupportedFxCodes(): Promise<string[]> {
  return Object.keys(await fetchFxRates());
}

async function loadPersisted(): Promise<Record<string, number>> {
  if (_cache) return _cache.rates;
  try {
    const raw = await AsyncStorage.getItem(CACHE_KEY);
    if (raw) {
      const c = JSON.parse(raw) as FxCache;
      if (c?.rates) { _cache = c; return c.rates; }
    }
  } catch { /* ignore */ }
  return {};
}

File: app-ios/VelaWallet/VelaWallet/Core/FiatFx.swift
//
//  FiatFx.swift
//  VelaWallet
//
//  USD → fiat, from the endpoint the person can change.
//
//  Ported from `app-web/vela-wallet/src/lib/services/fiat-fx.ts`. The default
//  is Vela's own Frankfurter instance (FOSS, no key, ~160 currencies including
//  VND); `设置 → 端点 → 汇率` replaces it, and `network_admin` owns that field.
//
//  The waterfall's SECOND rung: `FiatRates` (Chainlink) is asked first for the
//  sixteen currencies it can price, and this answers for everything else.
//
//  ## Two response shapes, because the provider is swappable
//
//  `normalize` accepts Frankfurter's array and the `{ rates: {...} }` object
//  both — which is what makes the endpoint genuinely configurable rather than
//  configurable-if-you-pick-the-same-vendor.
//
//  ## `vela.fxRates.v1` keeps the endpoint it came from
//
//  Web's field names and its six-hour TTL (FR-005). The URL rides in the cache
//  entry so changing the endpoint refetches instead of quietly serving the old
//  provider's numbers under the new one's name.
//

import Foundation

@MainActor
final class FiatFx {

    /// `network_admin`'s own default (`DEFAULT_FIAT_RATES_URL`), mirrored for
    /// the one case the shell needs before it has read the store.
    static let defaultURL = "https://vela-currency.getvela.app/v2/rates?base=USD"

    private static let cacheKey = VelaStore.Key.fxRates
    /// ECB/Frankfurter publish once per business day; six hours is plenty.
    private static let ttlMs: Double = 6 * 60 * 60 * 1000

    private let store: VelaStore
    private let accounts: AccountStore
    private var cached: (url: String, rates: [String: Double], atMs: Double)?

    init(store: VelaStore, accounts: AccountStore) {
        self.store = store
        self.accounts = accounts
    }

    /// USD → `code`, or `nil` when the endpoint cannot price it.
    func rate(_ code: String) async -> Double? {
        let upper = code.uppercased()
        if upper == "USD" { return 1 }
        let rates = await allRates()
        guard let rate = rates[upper], rate > 0 else { return nil }
        return rate
    }

    /// The whole USD-based map. Cached six hours in memory, persisted for a
    /// cold or offline start.
    func allRates() async -> [String: Double] {
        let url = await endpoint()
        let now = Date().timeIntervalSince1970 * 1000
        if let cached, cached.url == url, now - cached.atMs < Self.ttlMs { return cached.rates }

        let body = await CoreHTTP.getJSON(url, timeout: CoreHTTP.Timeout.fiatRates)
        guard let rates = Self.normalize(body) else {
            print("[vela-wallet] fx: no usable rates from \(url)")
            return persisted(url: url)
        }
        cached = (url, rates, now)
        store.writeObject(Self.cacheKey, ["url": url, "rates": rates, "at": now])
        return rates
    }

    /// Which currencies this endpoint can quote — the provider-driven list the
    /// picker is meant to offer rather than a constant.
    func supportedCodes() async -> [String] {
        Array(await allRates().keys)
    }

    private func endpoint() async -> String {
        let stored = await accounts.loadServiceEndpoints()["fiatRatesURL"] as? String
        guard let stored, !stored.isEmpty else { return Self.defaultURL }
        return stored
    }

    /// A provider response → `{ USD: 1, X: rate }`, or `nil` when it carried no
    /// rate at all.
    ///
    /// `nil` rather than `{ USD: 1 }`: a map with only the base in it can price
    /// nothing, and returning it would cache an answer that is indistinguishable
    /// from a working endpoint whose rates all happened to be missing.
    nonisolated static func normalize(_ body: Any?) -> [String: Double]? {
        var out: [String: Double] = ["USD": 1]

        if let rows = body as? [[String: Any]] {
            // Frankfurter v2: [{ base, quote, rate }, …]
            for row in rows {
                guard let code = (row["quote"] as? String)?.uppercased(), !code.isEmpty,
                      let rate = number(row["rate"]), rate > 0
                else { continue }
                out[code] = rate
            }
        } else if let object = body as? [String: Any],
                  let rates = object["rates"] as? [String: Any] {
            // open.er-api / v1: { rates: { EUR: 0.92, … } }
            for (code, value) in rates {
                guard let rate = number(value), rate > 0 else { continue }
                out[code.uppercased()] = rate
            }
        }

        return out.count > 1 ? out : nil
    }

    /// A JSON number or a numeric string — providers send both, and a rate that
    /// arrived as `"0.92"` is still a rate.
    nonisolated private static func number(_ value: Any?) -> Double? {
        if let number = value as? NSNumber { return number.doubleValue }
        if let text = value as? String { return Double(text) }
        return nil
    }

    /// The last map that reached disk, if it came from this same endpoint.
    private func persisted(url: String) -> [String: Double] {
        if let cached, cached.url == url { return cached.rates }
        let stored = store.readObject(Self.cacheKey)
        guard stored["url"] as? String == url else { return [:] }
        let rates = FiatRates.decodeRates(stored["rates"])
        guard !rates.isEmpty else { return [:] }
        cached = (url, rates, (stored["at"] as? NSNumber)?.doubleValue ?? 0)
        return rates
    }
}

/// One rung of the display currency's rate waterfall.
///
/// Two conform: `FiatRates` (Chainlink's on-chain fiat feeds) and `FiatFx` (the
/// configurable endpoint). A rung that cannot price a code answers `nil` and
/// the next is asked — which is why neither of them needs to know it is in a
/// waterfall at all.
@MainActor
protocol FiatRateSource {
    /// USD → `code`, or `nil` when this source cannot state it.
    func rate(_ code: String) async -> Double?
}

extension FiatFx: FiatRateSource {}
extension FiatRates: FiatRateSource {}
